### DISCUSSION TITLE
Consider SIMD Everywhere for uniform intrinsics support

### DISCUSSION THREAD

**Discussion by gdevenyi** [zlib-ng/zlib-ng#862] [Ideas]
> SIMDe is a permissively-licensed (MIT) header-only library which provides fast, portable implementations of SIMD intrinsics for platforms which aren’t natively supported by the API in question.
>
> For example, with SIMDe you can use SSE on ARM, POWER, WebAssembly, or almost any platform with a C compiler. That includes, of course, x86 CPUs which don’t support the ISA extension in question (e.g., calling AVX-512F functions on a CPU which doesn’t natively support them).
>
> If the target natively supports the SIMD extension in question there is no performance penalty for using SIMDe. Otherwise, accelerated implementations, such as NEON on ARM, AltiVec on POWER, WASM SIMD on WebAssembly, etc., are used when available to provide good performance.

https://simd-everywhere.github.io/blog/announcements/release/2020/06/21/0.5.0-release.html

**Comment by mtl1979** [comment]
I actually tried similar approach with adler32 in the early days and performance was a lot worse than non-SIMD version on some processors.... This is because only some specific AMD processors (read: Bulldozer only) have necessary instructions and emulating them is expensive.

Now we use versions optimized for specific prosessor architectures and the performance is a lot better especially when using AVX and AVX2. Not many processors have AVX512, but some versions of it include same, but wider, instructions that for example AMD Bulldozer had (XOP, horizontal add using one input vector register)...

**Comment by gdevenyi** [comment]
Cool. I'm not technically savvy enough in this to evaluate it usefulness directly, but I did see it as a possible way to reduce the development burden at the zlib-ng level by streamlining the code. Perhaps at least there might be bits to copy for providing performant emulation of calls where some arches don't support.

**Comment by mtl1979** [comment]
Horizontal add can be emulated with multiply-add (when available) with multiplier input vector as all decimal 1... This is because multiply-add widens the vector lanes like horizontal add... It however doesn't make sense to reload the multiplier on every call, so the load must be done when entering the outer function, not inside the "replacement" function.

Another specific issue with adler32 is the second multiplier vector which is basically each lane having one higher value than the previous lane.... In some processors (notably big-endian processors), there is own instruction to "load" internally the values for the vector lanes instead of loading them from memory.
